### DISCUSSION TITLE
View bindings with programmatic navigation links in SwiftUI

### DISCUSSION THREAD

**Discussion by thebarndog** [pointfreeco/swift-composable-architecture#996] [Q&A] [answered]
<img width="1182" alt="Screen Shot 2022-02-05 at 7 18 25 PM" src="https://user-images.githubusercontent.com/3139429/152666480-b3c8e3c1-a869-4855-ab86-79843938cc84.png">

Hello all! I have an interesting challenge around programatic navigation links, similar to the picture above. Sometimes, you want navigation to happen as a result of a state change, rather than from pressing on the link itself. In SwiftUI this is easy; you can just create a nav link with no title that's hidden. The problem is, because there's no link that the user can tap, the `set` closure never gets fed a true value. This problem also arises if you're using a custom back button.

I think the solution is easy but I wanted to get the core teams take on it. In this situation, I would just move those actions in the `set` to be side effects that are triggered by other actions, like saving the phrase or a navigation unwind action and use just a regular `Binding` or even `Binding.constant` since the view will recompute with the new value anyway. 

Do you have any suggestions on a better way to handle programmatic navigation links?

**Comment by stephencelis** [comment, marked as answer]
@thebarndog This seems like the general way to go about it, though it doesn't need to be driven by effects necessarily. As long as an action can mutate the state that determines if the link's binding is active or not, you will programmatically drive navigation.

In vanilla SwiftUI, this is an observable object mutating published state associated with that binding.

And in TCA, this is a reducer mutating state associated with the binding :smile:
